daemon/system-info.sh

```diff
@@ -323,6 +318,16 @@ elif [ "${KERNEL_NAME}" = FreeBSD ] ; then
         DISK_SIZE="$((total * 1024))"
 else
         if [ -d /sys/block ] ; then
+                # List of device majors we actually count towards total disk space.
+                # The meanings of these can be found in `Documentation/admin-guide/devices.txt` in the Linux sources.
+                # The ':' surrounding each number are important for matching.
+                dev_major_whitelist=':3:8:9:21:22:28:31:33:34:44:45:47:48:49:50:51:52:53:54:55:56:57:65:66:67:68:69:70:71:72:73:74:75:76:77:78:79:88:89:90:91:93:94:96:98:101:104:105:106:107:108:109:110:111:112:114:116:128:129:130:131:132:134:135:136:137:138:139:140:141:142:143:153:160:161:179:180:202:256:257:'
+
+                if [ "${VIRTUALIZATION}" != "unknown" ] ; then
```

Comment:
I don't think it's about CPU virtualization. Most virtual block device drivers would fall in this range, such as device mapper devices, which also typically use 253. The device mapper covers a huge range of configurations.

Reply:
We don't want DM devices though, we want the physical disks for the system. The point of looking at `/sys` is so we can match only on actual local physical devices that aren't removable storage and, which will, outside of this special case for virtualized systems, give us a 100% accurate listing of the real total disk space on the system (barring insanity from things like a host-protected area on ATA devices).

Reply:
Multipath I/O is one example, where the SCSI layer shows 8 identical devices but they are merged into 1 device-mapper multipath device, and the usable disk space is indeed 1/8th of what SCSI will show. It doesn't even have to be iSCSI devices, they can be attached via fibre channel which is effectively local storage.

Reply:
OK, let's approach my reasoning for this from a different direction:

We quite simply cannot make this perfect, there is literally no way to do so.

What's present here does a good enough job in a vast majority of cases and provides better accuracy for the common cases than for the uncommon ones.

More specifically:

* `df` is inaccurate (significantly so in some cases) in a number of very common configurations because it only looks at mounted devices. It reports low on a vast majority of 'trivial' storage configurations by anywhere from a few MB to double-digit percentages of total disk space, and high by arbitrary numbers in a number of non-trivial but reasonably common configurations, and that's all with it explicitly ignoring network and in-memory filesystems. It also reports inaccurately by some measures when dealing with network disks (it's inconsistent whether any arbitrary person would consider an iSCSI/NBD/ATAoE device to be a part of the system or not, and often depends on the exact circumstances).
* Checking `/sys` is extremely (read 'perfectly') accurate in a significant percentage of cases that `df` gives bogus results, because it looks at block devices directly. As it's implemented here, it correctly handles almost all standard cases except multipath, stuff that's thinly-provisioned at the hardware level, and VirtIO block devices (notably, it does handle VirtIO SCSI devices correctly, and that's what most people _should_ be using with QEMU/KVM), as well as covering a number of less common cases that `df` gets wrong (like thinly-provisioned logical volumes).

I would be more than happy to improve upon this logic, but there are two caveats:

* We need this functionality now for the ACLK stuff. While this is not perfect, it's good enough for the testing and pre-release stuff we're currently doing there, and there's no reason it can't be further improved later, once it's not potentially blocking development.
* Complexity of this code will increase _very_ quickly as we handle more and more specific use cases. Just off the top of my head, we're looking at at least a dozen more lines _just_ to properly handle the multipath case because we need to filter on DM table type (we _can't_ include all DM devices there, we'll get ridiculously high estimates if we do (potentially more than twice the actual disk space, and that's if they're not using thin provisioning, caching, or other complicated stacked layouts)). I'm arguably not even handling the VirtIO block case 'correctly' here because I'm not excluding other device types from the list.

If so desired, I can redo the sysfs handling to be far more robust, but that's probably going to take multiple days if we want to account for proper testing.